File: lua/lunest/bridge.lua
local M = {}

do
    ---@type string
    local ROOT_DIR
    ---@type { name: string, path: string }[]
    local TARGET_FILES
    ---@type string
    local MSG_FILE
    ---@type string?
    local INIT_FILE
    ---@type integer
    local TERM_WIDTH

    ---@return string
    function M.root_dir()
        return ROOT_DIR
    end
    ---@return { name: string, path: string }[]
    function M.get_target_files()
        return TARGET_FILES
    end
    ---@return string
    function M.get_msg_file_path()
        return MSG_FILE
    end
    ---@return string?
    function M.get_init_file()
        return INIT_FILE
    end
    ---@return integer
    function M.get_term_width()
        return TERM_WIDTH
    end
end

local json = {}
do
    ---@param obj any
    ---@return string
    local function format_q(obj)
        return (
            ("%q")
                :format(obj)
                :gsub("\\\n", [[\n]])
                :gsub([[([^\])\8]], [[%1\b]])
                :gsub([[([^\])\9]], [[%1\t]])
                :gsub([[([^\])\12]], [[%1\f]])
                :gsub([[([^\])\13]], [[%1\r]])
        )
    end

    ---@return string
    function json.encode(obj)
        local t = type(obj)
        if t == "nil" then
            return "null"
        elseif t == "number" or t == "boolean" then
            return tostring(obj)
        elseif t == "string" then
            return format_q(obj)
        elseif t ~= "table" then
            error(("invalid type '%s'"):format(t))
        elseif obj[1] then
            local ss = {}
            for _, value in ipairs(obj) do
                table.insert(ss, json.encode(value))
            end
            return "[" .. table.concat(ss, ",") .. "]"
        else
            local ss = {}
            for key, value in pairs(obj) do
                table.insert(ss, format_q(key) .. ":" .. json.encode(value))
            end
            return "{" .. table.concat(ss, ",") .. "}"
        end
    end
end

---@class lunest.bridge.Message
---@field TestStarted? lunest.bridge.TestStarted
---@field TestFinished? lunest.bridge.TestFinished

---@class lunest.bridge.TestStarted
---@field title string[]

---@class lunest.bridge.TestFinished
---@field title string[]
---@field error lunest.bridge.TestError?

---@class lunest.bridge.TestError
---@field message string
---@field traceback string
---@field info lunest.bridge.TestErrorInfo?

---@class lunest.bridge.TestErrorInfo
---@field Diff? { left: string, right: string }

---@type file*
local msg_file

---@param message lunest.bridge.Message
local function write_msg(message)
    if not msg_file then
        msg_file = assert(io.open(M.get_msg_file_path(), "a"))
    end
    assert(msg_file:write(json.encode(message) .. "\n"))
    msg_file:flush()
end

---@param title string[]
function M.start_test(title)
    write_msg({
        TestStarted = { title = title },
    })
end

---@param title string[]
---@param err lunest.bridge.TestError?
function M.finish_test(title, err)
    write_msg({
        TestFinished = {
            title = title,
            error = err,
        },
    })
end

function M.finish()
    if msg_file then
        assert(msg_file:close())
    end
end

return M

File: cli/src/bridge.rs
use std::{
    fmt, fs,
    io::{BufRead, BufReader},
    path::{Path, PathBuf},
};

use anyhow::{Context, Result};
use crossterm::{style::Stylize, terminal};
use serde::Deserialize;

#[derive(Debug)]
pub struct Bridge {
    path: PathBuf,
    reader: BufReader<fs::File>,
    line: String,
}

impl Bridge {
    pub fn new(temp_dir: &Path) -> Result<Self> {
        let path = temp_dir.join("messages.jsonl");
        fs::write(&path, "")?;
        let reader = BufReader::new(fs::File::open(&path)?);
        Ok(Self {
            path,
            reader,
            line: String::new(),
        })
    }

    pub fn read(&mut self) -> Result<Option<Message>> {
        while !self.line.ends_with('\n') {
            if self.reader.read_line(&mut self.line)? == 0 {
                return Ok(None);
            }
        }
        let msg: Message = {
            let line = self.line.trim_end_matches('\n');
            serde_json::from_str(line)
                .with_context(|| format!("failed to deserialize json: {line}"))?
        };
        self.line.clear();
        Ok(Some(msg))
    }

    pub fn overwrite_main_lua(
        &self,
        contents: &str,
        root_dir: impl AsRef<Path>,
        target_files: impl IntoIterator<Item = impl AsRef<Path>>,
        init_file: Option<impl AsRef<Path>>,
    ) -> String {
        let root_dir = root_dir.as_ref();
        let files: String = target_files.into_iter().fold(String::new(), |acc, p| {
            let p = p.as_ref();
            let name = p.strip_prefix(root_dir).unwrap_or(p);
            format!(
                "{acc}{{ name = \"{}\", path = \"{}\" }}, ",
                name.display().to_string().replace('\\', "/"),
                p.display().to_string().replace('\\', r"\\"),
            )
        });
        let contents = contents
            .replace(
                "--[[@replace = lunest.ROOT_DIR]]",
                &format!(
                    "= \"{}\"",
                    root_dir.display().to_string().replace('\\', r"\\")
                ),
            )
            .replace(
                "--[[@replace = lunest.TARGET_FILES]]",
                &format!("= {{ {files} }}"),
            )
            .replace(
                "--[[@replace = lunest.MSG_FILE]]",
                &format!(
                    "= \"{}\"",
                    self.path.display().to_string().replace('\\', r"\\")
                ),
            )
            .replace(
                "--[[@replace = lunest.TERM_WIDTH]]",
                &format!(
                    "= {}",
                    crossterm::terminal::size().map_or(60, |size| size.0)
                ),
            );
        if let Some(path) = init_file {
            contents.replace(
                "--[[@replace = lunest.INIT_FILE]]",
                &format!(
                    "= \"{}\"",
                    path.as_ref().display().to_string().replace('\\', r"\\"),
                ),
            )
        } else {
            contents
        }
    }
}

fn fmt_title(title: &[String]) -> String {
    title.join(&" :: ".grey().to_string())
}

#[derive(Debug, Deserialize)]
pub enum Message {
    TestStarted(TestStarted),
    TestFinished(TestFinished),
}

#[derive(Debug, Deserialize)]
pub struct TestStarted {
    title: Vec<String>,
}

impl fmt::Display for TestStarted {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{} {}",
            fmt_title(&self.title),
            ":".grey(),
            "RUNNING".cyan().bold()
        )
    }
}

#[derive(Debug, Deserialize)]
pub struct TestFinished {
    title: Vec<String>,
    error: Option<TestError>,
}

impl TestFinished {
    pub fn success(&self) -> bool {
        self.error.is_none()
    }
}

impl fmt::Display for TestFinished {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", terminal::Clear(terminal::ClearType::UntilNewLine))?;
        write!(f, "{}{} ", fmt_title(&self.title), ":".grey())?;
        if let Some(err) = &self.error {
            write!(f, "{}\n{}", "ERR".red().bold(), err)
        } else {
            write!(f, "{}", "OK".green().bold())
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct TestError {
    message: String,
    traceback: String,
    info: Option<TestErrorInfo>,
}

#[derive(Debug, Deserialize)]
pub enum TestErrorInfo {
    Diff { left: String, right: String },
}

impl fmt::Display for TestError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}\n", self.message.as_str().bold())?;
        if let Some(info) = &self.info {
            writeln!(f, "{info}")?;
        }
        writeln!(f, "{}:", "  stack traceback".bold())?;
        writeln!(f, "{}", self.traceback)?;
        Ok(())
    }
}

impl fmt::Display for TestErrorInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self {
            TestErrorInfo::Diff { left, right } => {
                writeln!(
                    f,
                    "{} ({} {} {}):",
                    "  difference".bold(),
                    "-left".red(),
                    "/".grey(),
                    "+right".green()
                )?;
                let delete = "-".red();
                let insert = "+".green();
                let diff = similar::TextDiff::from_lines(left, right);
                for change in diff.iter_all_changes() {
                    let line = change.value();
                    use similar::ChangeTag::*;
                    match change.tag() {
                        Equal => write!(f, " {}", line.grey())?,
                        Delete => write!(f, "{delete}{}", line.red())?,
                        Insert => write!(f, "{insert}{}", line.green())?,
                    }
                    if !line.ends_with('\n') {
                        writeln!(f)?;
                    }
                }
            }
        }
        Ok(())
    }
}
